openshift/dynamic/client.py

```diff
@@ -586,18 +490,276 @@ def to_dict(self):
+                    preferred = version_raw == group['preferredVersion']
+                    resources = {}
+                    if request_resources:
+                        resources = self.get_resources_for_api_version(prefix, group['name'], version, preferred)
+                    new_group[version] = ResourceGroup(preferred, resources=resources)
+                groups[prefix][group['name']] = new_group
+            self._cache['resources'] = groups
+        return self._cache['resources']
+
+    def _load_server_info(self):
+        if not self._cache.get('version'):
+            self._cache['version'] = {'kubernetes': load_json(self.client.request('get', '/version'))}
+            try:
+                self._cache['version']['openshift'] = load_json(self.client.request('get', '/version/openshift'))
+            except ApiException:
+                pass
+        self.__version = self._cache['version']
+
+    def get_resources_for_api_version(self, prefix, group, version, preferred):
+        """ returns a dictionary of resources associated with provided (prefix, group, version)"""
+
+        resources = {}
+        subresources = {}
+
+        path = '/'.join(filter(None, [prefix, group, version]))
+        resources_response = load_json(self.client.request('GET', path))['resources']
+
+        resources_raw = list(filter(lambda resource: '/' not in resource['name'], resources_response))
+        subresources_raw = list(filter(lambda resource: '/' in resource['name'], resources_response))
+        for subresource in subresources_raw:
+            resource, name = subresource['name'].split('/')
+            if not subresources.get(resource):
+                subresources[resource] = {}
+            subresources[resource][name] = subresource
+
+        for resource in resources_raw:
+            # Prevent duplicate keys
+            for key in ('prefix', 'group', 'api_version', 'client', 'preferred'):
+                resource.pop(key, None)
+
+            resources[resource['kind']] = Resource(
+                prefix=prefix,
+                group=group,
+                api_version=version,
+                client=self.client,
+                preferred=preferred,
+                subresources=subresources.get(resource['name']),
+                **resource)
+            resources['{}List'.format(resource['kind'])] = ResourceList(resources[resource['kind']])
+        return resources
+
+
+class LazyDiscoverer(Discoverer):
     """ A convenient container for storing discovered API resources. Allows
-        easy searching and retrieval of specific resources
+        easy searching and retrieval of specific resources.
+
+        Resources for the cluster are loaded lazily.
     """
 
-    def __init__(self, resources, client=None):
-        self.__resources = resources
-        self.__client = client
+    # Special key used to mark when cache needs to be upated
+    UPDATE_KEY = "__needs_update__"
```

Comment:
It might be cleaner to make an attribute on your discoverer like `__update_cache` and then override the `_write_cache` method to be something like:

```python
if self.__update_cache:
    super(LazyDiscoverer, self)._write_cache()
    self.__update_cache = False
```
Then you can get rid of most of the checks below and just set `self.__update_cache = True` whenever you discover a new resource, and just call `_write_cache` at the end.